### PR TITLE
Multiple validations of gTTA'ed tips, with delay

### PR DESCRIPTION
**Pull request by karimodm**
- Attempt up to `validationAttempts` to validate gTTA'ed tips
- Wait `validationDelay` before attempting validation against validators, to give them a timeframe to solidify tips

### CODE REVIEW

File: compass/conf/CoordinatorConfiguration.java [hunk 1]
```diff
@@ -56,6 +56,12 @@
   @Parameter(names = "-validator", description = "Validator nodes to use")
   public List<String> validators = new ArrayList<>();
 
+  @Parameter(names = "-validationAttempts", description = "If tips validation fails, obtain new tips and validate up to this number of attempts.")
+  public int validationAttempts = 10;
+
+  @Parameter(names = "-validationDelay", description = "Obtained tips might not be solid right away on the validators, sleep the specified amount of seconds before validating.")
```
Comment (GalRogozinski) [inline compass/conf/CoordinatorConfiguration.java:62]: is it seconds? milliseconds?
A better description please

File: compass/Coordinator.java [hunk 2]
```diff
@@ -293,7 +294,14 @@ private void start() throws InterruptedException {
...
         if (!validateTransactionsToApprove(trunk, branch)) {
-          throw new RuntimeException("Trunk & branch were not consistent!!! T: " + trunk + " B: " + branch);
+          if (++validationRetries >= config.validationAttempts) {
+            throw new RuntimeException("Trunk & branch were not consistent on multiple attempts!!! T: " + trunk + " B: " + branch);
+          } else {
+            // Perform gTTA and Validation again
+            continue;
+          }
+        } else {
+          validationRetries = 0;
```
Comment (GalRogozinski) [inline compass/Coordinator.java:304]: A question do we want to call GTTA again?
why not loop over the validation?

File: compass/conf/CoordinatorConfiguration.java (hunk 1 above)
Comment (karimodm) [reply compass/conf/CoordinatorConfiguration.java:62]: It's already written "sleep the specified amount of seconds before validating.". More specific than this?

File: compass/Coordinator.java (hunk 2 above)
Comment (karimodm) [reply compass/Coordinator.java:304]: We need gTTA again otherwise tips will become older and older with attempts.

**Comment by karimodm** [comment]
Fixes https://github.com/iotaledger/devops/issues/157

File: compass/conf/CoordinatorConfiguration.java (hunk 1 above)
Comment (GalRogozinski) [reply compass/conf/CoordinatorConfiguration.java:62]: couldn't see it on the github screen :+1:

File: compass/Coordinator.java [hunk 3]
```diff
@@ -292,10 +293,21 @@ private void start() throws InterruptedException {
...
+          if (!validateTransactionsToApprove(trunk, branch)) {
+            throw new RuntimeException("Trunk & branch not consistent");
+          }
+        } catch (RuntimeException e) {
+          if (++validationRetries >= config.validationAttempts) {
+            throw new RuntimeException("Trunk & branch were not consistent on multiple attempts!!! T: " + trunk + " B: " + branch);
+          } else {
+            log.warn("Validation failed, #{} of #{} attempts. Trying again...", validationRetries, config.validationAttempts);
+            // Perform gTTA and Validation again
+            continue;
```
Comment (GalRogozinski) [inline compass/Coordinator.java:307]: what was the problem you saw exactly before and how does this solve it?

Comment (karimodm) [reply compass/Coordinator.java:307]: Tips were obtain by gTTA, validated right away against the validator and if they were not found solid a RunTimeException was thrown.
I've discovered that IRI, unlike Hornet, make jota throw an exception if the transaction you are testing for consistency is missing from the DB; hence why I've also accounted for exceptions to be thrown by `validateTransactionsToApprove`.
In both cases this PR will make Compass do gTTA, wait a number of seconds and request validation against all configured validators. If any API or validation fail the entire process (gTTA + delay + validations) is performed all over again, up to a certain amount of retries; if the process goes well once the counter is reset, if the maximum number of attempts is reached Compass will throw the same way as it used to do before after a single failure.

**Pull request merged**